go/vt/topo/k8stopo/watchers.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package k8stopo implements topo.Server with the Kubernetes API as the backend.
+
+We expect the following behavior from the kubernetes client library:
+
+  - TODO
+
+We follow these conventions within this package:
+
+  - TODO
+*/
+package k8stopo
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+
+	vtv1beta1 "vitess.io/vitess/go/vt/topo/k8stopo/apis/topo/v1beta1"
+)
+
+// watcher is an interface that can be implemented to watch for changes to topology
+type watcher interface {
+	// onAddUpdate is called when a node is updated or added
+	// we do not differentiate between the two at this time
+	onAddUpdate(vtn *vtv1beta1.VitessTopoNode)
+
+	// onDelete is called when a node is deleted
+	onDelete(vtn *vtv1beta1.VitessTopoNode)
+}
+
+type watchGroup struct {
+	watchers *sync.Map
+}
+
+var _ watcher = (*watchGroup)(nil)
+
+// addWatcher adds the given watcher to the topo server watch map
+// it returns a function that can be called to remove the watcher
+func (wg *watchGroup) addWatcher(w watcher) func() {
+	// add to the watchers
+	id := uuid.NewString()
+
+	wg.watchers.Store(id, w)
+
+	return func() {
+		wg.watchers.Delete(id)
+	}
+}
```

Comment:
This is not a great use for a `sync.Map`. I think `watchGroup` should be modeled as a `map[watcher]struct{}` with an adjacent `sync.Mutex`. It has the same behavior as the `sync.Map`, but you don't need to use UUIDs or return a `func()` for cleanup (you can `delete(wg, w)` directly), you don't need to use `interface{}` all over the place, etc.

Generally speaking, `sync.Map` is not a very good primitive!